### PR TITLE
refactor: procedure storage

### PR DESCRIPTION
**Pull request by ZuLiangWang**
## Rationale
The current design of the procedure storage module is unreasonable. The data will be stored permanently and will not be cleared. This will cause the storage space of `ETCD` to be filled up quickly. We need to redesign the interface of procedure storage.

## Detailed Changes
* Add `ttl` config for `CreateOrUpdate`.
* Add `Delete` interface, it used to delete history procedure data and current procedure data.

## Test Plan
Pass all unit tests and integration tests.

### CODE REVIEW

File: server/coordinator/procedure/storage.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ import (
 )
 
 type Write interface {
-	CreateOrUpdate(ctx context.Context, meta Meta) error
+	CreateOrUpdate(ctx context.Context, meta Meta, ttl int64) error
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/storage.go:24]: `ttl` is ambiguous for callers, how about put it as:
```suggestion
	CreateOrUpdate(ctx context.Context, meta Meta, ttlSec int64) error
```

File: server/coordinator/procedure/storage_impl.go [hunk 2]
```diff
@@ -49,31 +53,66 @@ func NewEtcdStorageImpl(client *clientv3.Client, rootPath string, clusterID uint
...
-	keyPath := e.generaNormalKeyPath(id)
+func (e EtcdStorageImpl) MarkDeleted(ctx context.Context, procedureType Typ, id uint64) error {
+	keyPath := e.generaNormalKeyPath(procedureType, id)
 	meta, err := etcdutil.Get(ctx, e.client, keyPath)
 	if err != nil {
 		return errors.WithMessage(err, "get meta failed")
 	}
 
-	deletedKeyPath := e.generaDeletedKeyPath(id)
+	deletedKeyPath := e.generaDeletedKeyPath(procedureType, id)
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/storage_impl.go:115]: Could you give an explanation about why the type of procedure is part of the key? It looks very weird.

Comment (ZuLiangWang) [reply server/coordinator/procedure/storage_impl.go:115]: Considering query performance, currently we can only query all procedures through `List`. Using `Type` as part of the key can effectively reduce the amount of data queried in `list`.

File: server/coordinator/procedure/storage.go [hunk 3]
```diff
@@ -21,7 +21,7 @@ import (
 )
 
 type Write interface {
-	CreateOrUpdate(ctx context.Context, meta Meta) error
+	CreateOrUpdate(ctx context.Context, meta Meta, ttlSec int64) error
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/storage.go:24]: How about adding a new method called `CreateOrUpdateWithTTL` and keep the `CreateOrUpdate` same as before? If so, there is no need to change the irrelevant codes and there is no need to use zero ttlSec as a sentinel for no ttl.

File: server/coordinator/procedure/storage_impl.go [hunk 4]
```diff
@@ -49,31 +53,66 @@ func NewEtcdStorageImpl(client *clientv3.Client, rootPath string, clusterID uint
...
 
 	if _, err = e.client.Do(ctx, opPut); err != nil {
 		return errors.WithMessage(err, "etcd put data failed")
 	}
+
+	return nil
+}
+
+// Delete will delete the specified procedure,
+// and try to delete its corresponding history procedure if it exists.
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/storage_impl.go:86]: The style of the comment seems weird, how about letting the formatter break the original one line into two lines?

File: server/coordinator/procedure/procedure.go [hunk 5]
```diff
@@ -87,7 +87,7 @@ type Procedure interface {
 // Info is used to provide immutable description procedure information.
 type Info struct {
 	ID    uint64
-	Typ   Typ
+	Typ   Kind
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/procedure.go:90]: Kind？

File: server/coordinator/procedure/storage.go [hunk 6]
```diff
@@ -22,17 +22,19 @@ import (
 
 type Write interface {
 	CreateOrUpdate(ctx context.Context, meta Meta) error
+	CreateOrUpdateWithTTL(ctx context.Context, meta Meta, ttlSec int64) error
 }
 
 type Meta struct {
 	ID      uint64
-	Typ     Typ
+	Typ     Kind
```
Comment (ShiKaiWi) [inline server/coordinator/procedure/storage.go:30]: Kind？

**Review by ShiKaiWi — APPROVED** [review]

**Pull request merged**